File: Systems/MagicSystems/SurgebindingSystem.swift
import Foundation
import SpriteKit

/// Système de Surgebinding — pouvoirs des Chevaliers Radieux alimentés par la Lumière d'orage
final class SurgebindingSystem {

    // MARK: - Surges disponibles par ordre

    enum Surge: String, CaseIterable {
        case adhesion      // Coller des choses ensemble, boucliers
        case gravitation   // Manipuler la gravité, vol
        case abrasion      // Friction — glisser, esquiver
        case progression   // Croissance — guérison
        case illumination  // Illusions
        case transformation // Changer la matière (Soulcasting)
        case tension       // Rigidifier, renforcer
    }

    func surgesForOrder(_ order: RadiantOrder) -> (Surge, Surge) {
        switch order {
        case .windrunner:  return (.adhesion, .gravitation)
        case .lightweaver: return (.illumination, .transformation)
        case .bondsmith:   return (.tension, .adhesion)
        case .edgedancer:  return (.abrasion, .progression)
        }
    }

    // MARK: - Utiliser une Surge

    struct SurgeResult {
        let success: Bool
        let description: String
        let damage: Int
        let healing: Int
        let statusEffect: StatusEffectType?
        let duration: TimeInterval
        let stormlightCost: Double
    }

    func useSurge(
        _ surge: Surge,
        champion: inout Champion,
        targetPosition: GridPosition?
    ) -> SurgeResult {
        let cost: Double = 15.0

        guard let stormlight = champion.stormlightAmount, stormlight >= cost else {
            return SurgeResult(success: false, description: "Pas assez de Lumière d'orage",
                               damage: 0, healing: 0, statusEffect: nil, duration: 0, stormlightCost: 0)
        }

        champion.stormlightAmount = stormlight - cost

        switch surge {
        case .gravitation:
            return SurgeResult(
                success: true,
                description: "Lashage ! La gravité se retourne",
                damage: 20 + champion.baseStats.spirit,
                healing: 0,
                statusEffect: .flying,
                duration: 6.0,
                stormlightCost: cost
            )

        case .adhesion:
            return SurgeResult(
                success: true,
                description: "Adhésion — bouclier de Lumière d'orage !",
                damage: 0, healing: 0,
                statusEffect: .shielded,
                duration: 8.0,
                stormlightCost: cost
            )

        case .abrasion:
            return SurgeResult(
                success: true,
                description: "Abrasion — glissement parfait !",
                damage: 0, healing: 0,
                statusEffect: .haste,
                duration: 5.0,
                stormlightCost: cost
            )

        case .progression:
            let healAmount = 30 + champion.baseStats.spirit * 2
            champion.currentHP = min(champion.maxHP, champion.currentHP + healAmount)
            return SurgeResult(
                success: true,
                description: "Progression — guérison Radiante !",
                damage: 0,
                healing: healAmount,
                statusEffect: .healing,
                duration: 3.0,
                stormlightCost: cost
            )

        case .illumination:
            return SurgeResult(
                success: true,
                description: "Illumination — illusion créée !",
                damage: 0, healing: 0,
                statusEffect: .decoy,
                duration: 10.0,
                stormlightCost: cost
            )

        case .transformation:
            return SurgeResult(
                success: true,
                description: "Transformation — la matière se reconfigure !",
                damage: 25 + champion.baseStats.spirit,
                healing: 0,
                statusEffect: .stunned,
                duration: 2.0,
                stormlightCost: cost * 1.5
            )

        case .tension:
            return SurgeResult(
                success: true,
                description: "Tension — le sol se rigidifie !",
                damage: 0, healing: 0,
                statusEffect: .slowed,
                duration: 6.0,
                stormlightCost: cost
            )
        }
    }

    // MARK: - Recharge de Lumière d'orage

    func rechargeFromHighstorm(champion: inout Champion) {
        champion.stormlightAmount = Double(champion.maxInvestiture)
    }

    func rechargeFromGem(champion: inout Champion, gemSize: GemSize) {
        let amount: Double
        switch gemSize {
        case .chip:   amount = 10
        case .mark:   amount = 30
        case .broam:  amount = 75
        }
        champion.stormlightAmount = min(
            Double(champion.maxInvestiture),
            (champion.stormlightAmount ?? 0) + amount
        )
    }

    // MARK: - Effets visuels

    func gravitationLashEffect() -> SKAction {
        SKAction.sequence([
            SKAction.moveBy(x: 0, y: 50, duration: 0.3),
            SKAction.wait(forDuration: 5.4),
            SKAction.moveBy(x: 0, y: -50, duration: 0.3)
        ])
    }

    func stormlightGlowEffect() -> SKAction {
        let glow = SKAction.repeatForever(SKAction.sequence([
            SKAction.colorize(with: .cyan, colorBlendFactor: 0.4, duration: 0.5),
            SKAction.colorize(withColorBlendFactor: 0.1, duration: 0.5)
        ]))
        return glow
    }
}

// MARK: - Gem Types

enum GemSize: String, Codable {
    case chip   // Petite gemme
    case mark   // Moyenne
    case broam  // Grande (précieuse)
}

File: cosmere-web/src/game/magic/SurgebindingSystem.ts
import { RadiantOrder } from '@/data/types';
import { MagicSystem, MagicResult, emptyResult } from './MagicResult';

const STORMLIGHT_COST = 15;
const TRANSFORMATION_MULTIPLIER = 1.5;

interface SurgeInfo {
  name: string;
  description: string;
}

const SURGES: Record<string, SurgeInfo> = {
  adhesion: { name: 'Adhésion', description: 'Crée un bouclier protecteur' },
  gravitation: { name: 'Gravitation', description: 'Inverse la gravité, permet le vol' },
  abrasion: { name: 'Abrasion', description: 'Réduit la friction, vitesse accrue' },
  progression: { name: 'Progression', description: 'Soigne les blessures' },
  illumination: { name: 'Illumination', description: 'Crée des leurres illusoires' },
  transformation: { name: 'Transformation', description: 'Modifie la matière (Soulcasting)' },
  tension: { name: 'Tension', description: 'Rigidifie le sol, ralentit les ennemis' },
};

const ORDER_SURGES: Record<RadiantOrder, [string, string]> = {
  windrunner: ['adhesion', 'gravitation'],
  lightweaver: ['illumination', 'transformation'],
  bondsmith: ['tension', 'adhesion'],
  edgedancer: ['abrasion', 'progression'],
};

export class SurgebindingSystem implements MagicSystem {
  readonly name = 'Lien des Tempêtes';
  readonly worldID = 'roshar';

  private order: RadiantOrder = 'windrunner';

  setOrder(order: RadiantOrder): void {
    this.order = order;
  }

  getOrder(): RadiantOrder {
    return this.order;
  }

  canUseAbility(abilityID: string, currentInvestiture: number): boolean {
    const surge = abilityID.replace('surge_', '');
    if (!SURGES[surge]) return false;
    const surges = ORDER_SURGES[this.order];
    if (!surges.includes(surge)) return false;
    const cost = surge === 'transformation' ? STORMLIGHT_COST * TRANSFORMATION_MULTIPLIER : STORMLIGHT_COST;
    return currentInvestiture >= cost;
  }

  useAbility(abilityID: string, spirit: number): MagicResult {
    const surge = abilityID.replace('surge_', '');
    const result = emptyResult();

    if (!SURGES[surge]) {
      result.description = 'Surge inconnue';
      return result;
    }

    const surges = ORDER_SURGES[this.order];
    if (!surges.includes(surge)) {
      result.description = `Les ${this.order}s ne maîtrisent pas ${SURGES[surge].name}`;
      return result;
    }

    result.success = true;
    result.resourceCost = surge === 'transformation'
      ? Math.ceil(STORMLIGHT_COST * TRANSFORMATION_MULTIPLIER)
      : STORMLIGHT_COST;
    result.resourceType = 'stormlight';

    switch (surge) {
      case 'adhesion':
        result.statusEffect = 'shielded';
        result.statusDuration = 8;
        result.description = 'Bouclier d\'Adhésion activé !';
        break;
      case 'gravitation':
        result.statusEffect = 'flying';
        result.statusDuration = 6;
        result.description = 'Gravitation inversée !';
        break;
      case 'abrasion':
        result.statusEffect = 'invisible'; // haste effect
        result.statusDuration = 5;
        result.description = 'Friction réduite — vitesse accrue !';
        break;
      case 'progression':
        result.healing = 30 + spirit * 2;
        result.statusEffect = 'healing';
        result.statusDuration = 3;
        result.description = 'Progression — soins appliqués';
        break;
      case 'illumination':
        result.statusEffect = 'invisible';
        result.statusDuration = 10;
        result.description = 'Illusion créée !';
        break;
      case 'transformation':
        result.damage = 25 + spirit;
        result.statusEffect = 'stunned';
        result.statusDuration = 2;
        result.description = 'Soulcasting — transformation !';
        break;
      case 'tension':
        result.statusEffect = 'slowed';
        result.statusDuration = 6;
        result.areaOfEffect = true;
        result.areaRadius = 100;
        result.description = 'Tension — sol rigidifié !';
        break;
    }

    return result;
  }

  getAvailableAbilities(): string[] {
    return ORDER_SURGES[this.order].map(s => `surge_${s}`);
  }

  rechargeFromGem(gemType: 'chip' | 'mark' | 'broam'): number {
    const amounts = { chip: 10, mark: 30, broam: 75 };
    return amounts[gemType];
  }
}
